File: macos/Sources/Claudario/Install/HookInstaller.swift
import Foundation

enum HookInstallerError: LocalizedError {
    case hookScriptMissing
    case settingsCorrupt(String)
    case writeFailed(String)

    var errorDescription: String? {
        switch self {
        case .hookScriptMissing:
            return "Bundled hook script missing from app resources."
        case .settingsCorrupt(let msg):
            return "~/.claude/settings.json is not valid JSON: \(msg)"
        case .writeFailed(let msg):
            return "Failed to write settings: \(msg)"
        }
    }
}

final class HookInstaller {
    private static let hookEvents = [
        "UserPromptSubmit",
        "PreToolUse",
        "PostToolUse",
        "Stop",
        "SubagentStop",
        "Notification",
        "SessionStart",
        "SessionEnd",
        "PreCompact",
    ]
    private static let marker = "/.claudario/"

    private var home: URL { FileManager.default.homeDirectoryForCurrentUser }
    private var configDir: URL { home.appendingPathComponent(".claudario") }
    private var hookDest: URL { configDir.appendingPathComponent("hook") }
    private var claudeDir: URL { home.appendingPathComponent(".claude") }
    private var settingsPath: URL { claudeDir.appendingPathComponent("settings.json") }

    @discardableResult
    func install() throws -> URL {
        try FileManager.default.createDirectory(at: configDir, withIntermediateDirectories: true)
        try copyBundledHook()

        try FileManager.default.createDirectory(at: claudeDir, withIntermediateDirectories: true)
        var json: [String: Any] = [:]
        if FileManager.default.fileExists(atPath: settingsPath.path) {
            let data = try Data(contentsOf: settingsPath)
            let backup = settingsPath
                .deletingPathExtension()
                .appendingPathExtension("json.bak.\(Int(Date().timeIntervalSince1970))")
            try data.write(to: backup)
            do {
                if let parsed = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    json = parsed
                }
            } catch {
                throw HookInstallerError.settingsCorrupt(error.localizedDescription)
            }
        }

        var hooks = (json["hooks"] as? [String: Any]) ?? [:]
        let entry: [String: Any] = [
            "matcher": "*",
            "hooks": [["type": "command", "command": hookDest.path]],
        ]
        for event in Self.hookEvents {
            var arr = (hooks[event] as? [[String: Any]]) ?? []
            arr.removeAll { Self.entryIsClaudario($0) }
            arr.append(entry)
            hooks[event] = arr
        }
        json["hooks"] = hooks

        do {
            let out = try JSONSerialization.data(
                withJSONObject: json, options: [.prettyPrinted, .sortedKeys])
            try out.write(to: settingsPath)
        } catch {
            throw HookInstallerError.writeFailed(error.localizedDescription)
        }
        return settingsPath
    }

    func uninstall() throws {
        guard FileManager.default.fileExists(atPath: settingsPath.path) else { return }
        let data = try Data(contentsOf: settingsPath)
        guard var json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { return }
        guard var hooks = json["hooks"] as? [String: Any] else { return }

        for (event, value) in hooks {
            guard var arr = value as? [[String: Any]] else { continue }
            arr.removeAll { Self.entryIsClaudario($0) }
            if arr.isEmpty {
                hooks.removeValue(forKey: event)
            } else {
                hooks[event] = arr
            }
        }
        if hooks.isEmpty {
            json.removeValue(forKey: "hooks")
        } else {
            json["hooks"] = hooks
        }
        let out = try JSONSerialization.data(
            withJSONObject: json, options: [.prettyPrinted, .sortedKeys])
        try out.write(to: settingsPath)
    }

    private func copyBundledHook() throws {
        guard let bundled = Bundle.main.url(forResource: "claudario-hook", withExtension: nil) else {
            throw HookInstallerError.hookScriptMissing
        }
        if FileManager.default.fileExists(atPath: hookDest.path) {
            try FileManager.default.removeItem(at: hookDest)
        }
        try FileManager.default.copyItem(at: bundled, to: hookDest)
        try FileManager.default.setAttributes([.posixPermissions: 0o755], ofItemAtPath: hookDest.path)
    }

    private static func entryIsClaudario(_ entry: [String: Any]) -> Bool {
        guard let inner = entry["hooks"] as? [[String: Any]] else { return false }
        return inner.contains { ($0["command"] as? String)?.contains(Self.marker) == true }
    }
}

File: windows/Claudario.Windows/Install/HookInstaller.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Claudario.Windows.Install;

/// <summary>
/// Installs / uninstalls the Claudario hook entries in
/// %USERPROFILE%\.claude\settings.json.  Matches the macOS HookInstaller
/// port 1-to-1 in behaviour; uses System.Text.Json.Nodes for mutable JSON.
/// </summary>
public sealed class HookInstaller
{
    private static readonly string[] HookEvents =
    [
        "UserPromptSubmit", "PreToolUse", "PostToolUse",
        "Stop", "SubagentStop", "Notification",
        "SessionStart", "SessionEnd", "PreCompact",
    ];

    // Substring present in any command we wrote — used to identify our entries.
    private const string Marker = ".claudario";

    private string Home        => Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
    private string ConfigDir   => Path.Combine(Home, ".claudario");
    private string HookDest    => Path.Combine(ConfigDir, "hook.cmd");
    private string ClaudeDir   => Path.Combine(Home, ".claude");
    private string SettingsPath => Path.Combine(ClaudeDir, "settings.json");

    // ── Install ───────────────────────────────────────────────────────────────

    public string Install()
    {
        Directory.CreateDirectory(ConfigDir);
        CopyBundledHook();

        Directory.CreateDirectory(ClaudeDir);

        JsonObject root = ReadOrCreateSettings();

        // Backup before modifying
        if (File.Exists(SettingsPath))
        {
            string bak = SettingsPath + $".bak.{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}";
            File.Copy(SettingsPath, bak, overwrite: true);
        }

        var hooksObj = root["hooks"]?.AsObject() ?? new JsonObject();
        root["hooks"] = hooksObj;

        // Command: invoke hook.cmd via cmd.exe so Claude Code needn't know about .cmd
        string cmd = $"cmd.exe /c \"{HookDest}\"";
        var entry = new JsonObject
        {
            ["matcher"] = "*",
            ["hooks"]   = new JsonArray
            {
                new JsonObject { ["type"] = "command", ["command"] = cmd }
            },
        };

        foreach (string ev in HookEvents)
        {
            var arr = hooksObj[ev]?.AsArray() ?? new JsonArray();
            // Remove any existing Claudario entries then append the fresh one
            for (int i = arr.Count - 1; i >= 0; i--)
                if (EntryIsClaudario(arr[i]?.AsObject())) arr.RemoveAt(i);
            arr.Add(entry.DeepClone());
            hooksObj[ev] = arr;
        }

        WriteSettings(root);
        return SettingsPath;
    }

    // ── Uninstall ─────────────────────────────────────────────────────────────

    public void Uninstall()
    {
        if (!File.Exists(SettingsPath)) return;
        JsonObject root = ReadOrCreateSettings();
        if (root["hooks"] is not JsonObject hooksObj) return;

        foreach (string ev in HookEvents)
        {
            if (hooksObj[ev] is not JsonArray arr) continue;
            for (int i = arr.Count - 1; i >= 0; i--)
                if (EntryIsClaudario(arr[i]?.AsObject())) arr.RemoveAt(i);
            if (arr.Count == 0) hooksObj.Remove(ev);
        }
        if (hooksObj.Count == 0) root.Remove("hooks");

        WriteSettings(root);
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void CopyBundledHook()
    {
        // Look next to the exe, or fall back to the working directory
        string src = Path.Combine(AppContext.BaseDirectory, "claudario-hook.cmd");
        if (!File.Exists(src))
            throw new FileNotFoundException(
                "claudario-hook.cmd not found next to the executable.", src);

        File.Copy(src, HookDest, overwrite: true);
    }

    private JsonObject ReadOrCreateSettings()
    {
        if (!File.Exists(SettingsPath)) return new JsonObject();
        try
        {
            var node = JsonNode.Parse(File.ReadAllText(SettingsPath));
            return node?.AsObject() ?? new JsonObject();
        }
        catch (JsonException ex)
        {
            throw new InvalidOperationException(
                $"~/.claude/settings.json is not valid JSON: {ex.Message}", ex);
        }
    }

    private void WriteSettings(JsonObject root)
    {
        var opts = new JsonSerializerOptions { WriteIndented = true };
        File.WriteAllText(SettingsPath, root.ToJsonString(opts));
    }

    private static bool EntryIsClaudario(JsonObject? entry)
    {
        if (entry?["hooks"] is not JsonArray inner) return false;
        foreach (var h in inner)
            if (h?["command"]?.GetValue<string>()?.Contains(Marker) == true) return true;
        return false;
    }
}
